src/proto/sim.rs

```diff
@@ -0,0 +1,1103 @@
+    }
+
+    /// Drains all queued events and returns them in a sorted vector.
+    pub fn events_sorted(&mut self) -> Vec<(PI, TopicId, Event<PI>)> {
+        sort(self.events().collect())
+    }
+
+    /// Returns all active connections.
+    pub fn conns(&self) -> Vec<(PI, PI)> {
+        sort(self.conns.iter().cloned().map(Into::into).collect())
+    }
+
+    /// Queues and performs a command.
+    pub fn command(&mut self, peer: PI, topic: TopicId, command: Command<PI>) {
+        debug!(?peer, "~~ COMMAND {command:?}");
+        self.queue
+            .insert(self.time, peer, InEvent::Command(topic, command));
+        self.tick();
+    }
+
+    pub fn peer_states(&self) -> impl Iterator<Item = &State<PI, R>> {
+        self.peers.values()
+    }
+
+    pub fn peer_ids(&self) -> impl Iterator<Item = PI> + '_ {
+        self.peers.keys().cloned()
+    }
+
+    pub fn current_tick(&self) -> u32 {
+        self.tick as u32
+    }
+
+    pub fn elapsed(&self) -> Duration {
+        self.time.duration_since(self.start)
+    }
+
+    pub fn elapsed_fmt(&self) -> String {
+        format!(
+            "{:>2.4}s",
+            self.time.duration_since(self.start).as_secs_f32()
+        )
+    }
+
+    /// Runs the simulation for `n` times the maximum latency between peers.
+    pub fn run_roundtrips(&mut self, n: usize) {
+        let duration = self.config.latency.max() * n as u32;
+        self.run_duration(duration)
+    }
+
+    /// Runs the simulation for `timeout`.
+    pub fn run_duration(&mut self, timeout: Duration) {
+        let end = self.time + timeout;
+        while self.queue.next_before(end) {
+            self.tick();
+        }
+        assert!(self.time <= end);
+        self.time = end;
+    }
+
+    /// Runs the simulation while `f` returns `true`.
+    ///
+    /// The callback will be called for each emitted event.
+    pub fn run_while(&mut self, mut f: impl FnMut(PI, TopicId, Event<PI>) -> bool) {
+        'outer: loop {
+            while let Some((peer, topic, event)) = self.events.pop_front() {
+                if !f(peer, topic, event) {
+                    break 'outer;
+                }
+            }
+            self.tick();
+        }
+    }
+
+    /// Runs the simulation while `f` returns `true`, aborting after `timeout`.
+    ///
+    /// The callback will be called for each emitted event.
+    pub fn run_while_with_timeout(
+        &mut self,
+        timeout: Duration,
+        mut f: impl FnMut(PI, TopicId, Event<PI>) -> bool,
+    ) {
+        let end = self.time + timeout;
+        loop {
+            while let Some((peer, topic, event)) = self.events.pop_front() {
+                if !f(peer, topic, event) {
+                    return;
+                }
+            }
+            if self.queue.next_before(end) {
+                self.tick();
+            } else {
+                break;
+            }
+        }
+        assert!(self.time <= end);
+        self.time = end;
```

Comment:
same question, why set the end time to an artificial end?

Reply:
see comment above